omnibus/files/private-chef-ctl-commands/cleanup_bifrost.rb

```diff
@@ -0,0 +1,213 @@
+    end
+  end
+end
+
+def safety_check(db)
+  res = db.exec("SELECT * FROM pg_tables
+                          WHERE tablename='cleanup_tracking_auth_actors'")
+  if res.ntuples > 0
+    puts "ERROR: cleanup_tracking_auth_actors already exists.  cleanup-bifrost may be running."
+    puts "ERROR: If you are sure cleanup-bifrost is not running, you can clean up the tracking tables with: chef-backend-ctl cleanup-bifrost --force-cleanup"
+  end
+end
+
+def install_bifrost_tracking_table(db)
+  puts "Installing tracking tables into bifrost database"
+  db.exec(CREATE_SQL)
+end
+
+def remove_bifrost_tracking_table(db)
+  puts "Removing tracking tables into bifrost database"
+  db.exec(DELETE_SQL)
+end
+
+def print_and_return_estimate(known_actor_list, db)
+  tcount = timed "Fetchings all actor counts" do
+    db.exec_params("SELECT count(*) FROM auth_actor").first['count'].to_i
+  end
+
+  estimated_del_count = [0, tcount - known_actor_list.length].max
+  puts "\n----------------------------------------"
+  puts " Total chef users+clients: #{known_actor_list.length}"
+  puts "Total bifrost auth_actors: #{tcount}"
+  puts "Deletion Candidates (est): #{estimated_del_count}"
+  puts "----------------------------------------\n"
+
+  estimated_del_count
+end
+
+def run_bifrost_scan(known_actor_list, batch_size, db)
+  total_deleted = 0
+  known_actors_str = to_pg_list(known_actor_list)
+  loop do
+    puts "Processing batch of #{batch_size}"
+
+    candidates = timed "    Retrieving batch from auth_actors" do
+      db.exec("SELECT authz_id FROM auth_actor WHERE authz_id NOT IN (#{known_actors_str}) LIMIT $1",
+              [batch_size])
+        .map {|r| r['authz_id'] }
+    end
+
+    if candidates.length == 0
+      puts "No more candidates for deletion."
+      break
+    end
+
+    confirmed_candidates = timed "    Ignoring any recently created actors from candidates (n = #{candidates.length})" do
+      sql = "SELECT authz_id FROM cleanup_tracking_auth_actors WHERE authz_id IN (#{to_pg_list(candidates)})"
+      newly_created = db.exec(sql).map {|r| r['authz_id'] }
+      ret = candidates - newly_created
+      printf " (ignored: #{newly_created.length}) (remaining: #{ret.length})"
+      ret
+    end
+
+    if confirmed_candidates.length == 0
+      puts "All candidates in batch were newly created. Stopping"
+      break
+    end
+
+    delete_res = timed "    Deleting batch of #{confirmed_candidates.length} unknown auth_actors" do
+      db.exec("DELETE FROM auth_actor WHERE authz_id IN (#{to_pg_list(confirmed_candidates)})")
+    end
+
+    total_deleted += delete_res.cmd_tuples
+    break if delete_res.cmd_tuples == 0
+  end
+  puts "Total auth_actors removed: #{total_deleted}"
+end
+
+def timed(description)
+  printf description
+  st = Time.now
+  res = yield
+  printf ": #{Time.now - st}\n"
+  res
+end
+
+def to_pg_list(array)
+  "'" + array.join("','") + "'"
+end
+
+# TODO(ssd) 2017-07-24: I know these functions need to use veil to get
```

Comment:
> so the user in question can install this by just dropping it into the correct directory.

Oh! I see... 💡 (I wonder if we need a "properer" mechanism for this... like, _"dump your scripts here"_, have some metadata in them so we've got some basic version checks... *this* is a little wild west...)